### ISSUE TITLE
Better error when the cert is expired

### ISSUE THREAD

**Issue by innnkkki** [GENI-NSF/geni-tools#264]
Omni can figure out locally if a cert/credential has expired and warn the user, 
before ever talking to the AM. 

Add a more user friendly way to handle this. 

 _Imported from trac ticket [#264](http://trac.gpolab.bbn.com/gcf/ticket/264),  created by nriga on 05-03-2013 at 13:40, last modified: 05-15-2015 at 08:54_

**Comment by innnkkki** [comment]
When I invoke omni with an expired cert and the --debug flag, omni barfs 
```
DEBUG:omni.protogeni:Traceback (most recent call last):
  File "/home/nriga/gcf/src/omnilib/util/dossl.py", line 69, in _do_ssl
    result = fn(*args)
  File "/usr/lib/python2.6/xmlrpclib.py", line 1199, in __call__
    return self.__send(self.__name, args)
  File "/usr/lib/python2.6/xmlrpclib.py", line 1489, in __request
    verbose=self.__verbose
  File "/usr/lib/python2.6/xmlrpclib.py", line 1235, in request
    self.send_content(h, request_body)
  File "/usr/lib/python2.6/xmlrpclib.py", line 1349, in send_content
    connection.endheaders()
  File "/usr/lib/python2.6/httplib.py", line 904, in endheaders
    self._send_output()
  File "/usr/lib/python2.6/httplib.py", line 776, in _send_output
    self.send(msg)
  File "/usr/lib/python2.6/httplib.py", line 735, in send
    self.connect()
  File "/usr/lib/python2.6/httplib.py", line 1112, in connect
    self.sock = ssl.wrap_socket(sock, self.key_file, self.cert_file)
  File "/usr/lib/python2.6/ssl.py", line 350, in wrap_socket
    suppress_ragged_eofs=suppress_ragged_eofs)
  File "/usr/lib/python2.6/ssl.py", line 118, in __init__
    self.do_handshake()
  File "/usr/lib/python2.6/ssl.py", line 293, in do_handshake
    self._sslobj.do_handshake()
SSLError: [Errno 1] _ssl.c:480: error:14094415:SSL routines:SSL3_READ_BYTES:sslv3 alert certificate expired
```
The execution continues but this case should be handled better. 

_Trac comment by nriga (github user: innnkkki) on 05-03-2013 at 14:12_

**Comment by ahelsing** [comment]
Omni has code that claims to handle exactly this case (dossl.py line 120 looks for ssl.SSLError errno 1 saying 'error:14094415' and tries to return a message telling you to go renew your cert. Apparently this isn't happening?

_Trac comment by ahelsing on 05-03-2013 at 14:26_

**Comment by innnkkki** [comment]
Looking at the output closer, somewhere it prints:
ERROR:omni.protogeni:Cannot create a PG slice without a valid user credential: Your user certificate urn:publicid:IDN+pgeni.gpolab.bbn.com+user+inki has expired at 20130416163637Z. ProtoGENI users should log in to their account at the ProtoGENI website at http://boss.pgeni.gpolab.bbn.com and create and download a new certificate. 

But OMNI does not exit after that and it is not part of the summary, so it is hard to parse. It is also probably too verbose. 

_Trac comment by nriga (github user: innnkkki) on 05-03-2013 at 14:30_

**Comment by ahelsing** [comment]
What shorter (less verbose) text would you like to see?

To get this part of the summary, the main amhandler (or whoever called framework.get_user_cred) probably needs to notice that this is a particularly bad error, not try other aggregates if any, and put this in the summary, and then exit. Getting that signal to propagate up may take some work. In particular, the framework methods generally don't return the message. Perhaps we should raise an OmniError here. Perhaps we need code like this wherever we call get_user_cred:
```
            if self.opts.devmode:
                self.logger.warn(msg + " ... but continuing")
                credxml = cred
            else:
                self._raise_omni_error(msg)
```

We don't want the underlying code to abruptly exit, in case this is called from a script. I think.



_Trac comment by ahelsing on 05-03-2013 at 14:59_

**Comment by ahelsing** [comment]
I've now added a handler_utils is_user_cert_expired method. Not used yet.

_Trac comment by ahelsing on 08-07-2013 at 10:06_

**Comment by ahelsing** [comment]
perhaps framework_base could call that utility method, raising an OmniError if the cert is expired. The oscript.load_framework would wrap the init in a try/except. And right there we can decide if we continue or what.

_Trac comment by ahelsing on 04-02-2014 at 13:42_

**Comment by ahelsing** [comment]
After Saturday I'll have a cert I can use to test this.

_Trac comment by ahelsing on 06-20-2014 at 16:13_
